src/serializer.cpp

```diff
@@ -144,6 +161,180 @@ prepare() ->
+
+            chunk_header = buf;
+        }
+
+        auto get_output = [&]() -> buffers::mutable_buffer
+        {
+            auto mbs = tmp0_.prepare(tmp0_.capacity());
+            auto buf = *mbs.begin();
+            if( buf.size() == 0 )
+            {
+                auto p = mbs.begin();
+                ++p;
+                buf = *p;
+            }
+
+            if( is_chunked_ )
+            {
+                if( buf.size() <
+                    detail::crlf_len_ + detail::last_chunk_len_ + 1 )
+                    return {};
+
+                auto n =
+                    buf.size() -
+                    detail::crlf_len_ -
+                    detail::last_chunk_len_;
+
+                buf = { buf.data(), n };
+            }
+
+            // BOOST_ASSERT(buf.size() > 0);
+            return buf;
+        };
+
+        auto get_input = [&]() -> buffers::const_buffer
+        {
+            if( st_ == style::buffers )
+            {
+                if( buffers::buffer_size(buf_) == 0 )
+                    return {};
+
+                auto buf = *(buf_.data());
+                BOOST_ASSERT(buf.size() > 0);
+                return buf;
+            }
+            else
+            {
+                if( zbuf.size() == 0 )
+                    return {};
+
+                auto cbs = zbuf.data();
+                auto buf = *cbs.begin();
+                if( buf.size() == 0 )
+                {
+                    auto p = cbs.begin();
+                    ++p;
+                    buf = *p;
+                }
+                BOOST_ASSERT(buf.size() > 0);
+                return buf;
+            }
+        };
+
+        std::size_t num_written = 0;
+        while( true )
+        {
+            auto in = get_input();
+            auto out = get_output();
+            if( out.size() < 6 + 1 )
+            {
+                BOOST_ASSERT(tmp0_.size() > 0);
+                break;
+            }
+
+            auto results = zlib_filter_->on_process(
+                out, in, more_);
+
+            if( st_ == style::buffers )
+            {
+                buf_.consume(results.in_bytes);
+                if( buffers::buffer_size(buf_) == 0 )
+                    more_ = false;
+            }
+            else
+                zbuf.consume(results.in_bytes);
+
+            if( results.out_bytes == 0 )
+                break;
+
+            BOOST_ASSERT(results.out_bytes > 0);
+
+            num_written += results.out_bytes;
+            tmp0_.commit(results.out_bytes);
+        }
+
+        BOOST_ASSERT(tmp0_.size() > 0);
+        if( is_chunked_ )
+        {
+            write_chunk_header(
+                chunk_header, num_written);
+
+            buffers::buffer_copy(
+                tmp0_.prepare(2),
+                buffers::const_buffer("\r\n", 2));
+            tmp0_.commit(2);
+
+            if( static_cast<zlib::zlib_filter*>(
+                    zlib_filter_)->is_done() )
```

Comment:
`zlib_filter_` should be an abstract interface coming from http_proto. It should not be referring to any symbols in http_proto_zlib. The cast should not be needed.